server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1589,6 +1509,158 @@ static String resolveExpression(String expression, LongSupplier getTime) {
+            if (indexAbstraction.getType() == Type.ALIAS && context.getOptions().ignoreAliases()) {
+                if (ignoreUnavailable) {
+                    return false;
+                } else {
+                    throw aliasesNotSupportedException(expression);
+                }
+            }
+            if (indexAbstraction.isDataStreamRelated() && context.includeDataStreams() == false) {
+                if (ignoreUnavailable) {
+                    return false;
+                } else {
+                    IndexNotFoundException infe = notFoundException(expression);
+                    // Allows callers to handle IndexNotFoundException differently based on whether data streams were excluded.
+                    infe.addMetadata(EXCLUDED_DATA_STREAMS_KEY, "true");
+                    throw infe;
+                }
+            }
+            return true;
+        }
+
+        private static void validateAliasOrIndex(ExpressionIterable.Expression expression) {
+            if (expression.isExclusion()) {
+                return;
+            }
+            if (Strings.isEmpty(expression.toString())) {
+                throw notFoundException(expression.toString());
+            }
+            // Expressions can not start with an underscore. This is reserved for APIs. If the check gets here, the API
+            // does not exist and the path is interpreted as an expression. If the expression begins with an underscore,
+            // throw a specific error that is different from the [[IndexNotFoundException]], which is typically thrown
+            // if the expression can't be found.
+            if (expression.toString().charAt(0) == '_') {
+                throw new InvalidIndexNameException(expression.toString(), "must not start with '_'.");
+            }
+        }
+    }
+
+    /**
+     * Used to iterate the expression list and work out which expression item is a wildcard or an exclusion.
+     */
+    public static final class ExpressionIterable implements Iterable<ExpressionIterable.Expression> {
+        private final List<String> expressions;
+        private final int indexOfFirstWildcard;
+
+        public final class Expression {
+            private final int idx;
+
+            Expression(int idx) {
+                assert idx >= 0;
+                assert idx < ExpressionIterable.this.expressions.size();
+                this.idx = idx;
+            }
+
+            @Override
+            public String toString() {
+                if (isExclusion()) {
+                    // drop the leading "-" if exclusion
+                    return get().substring(1);
+                } else {
+                    return get();
+                }
+            }
+
+            public boolean isWildcard() {
+                return idx >= ExpressionIterable.this.indexOfFirstWildcard && IndexNameExpressionResolver.isWildcard(get());
```

Comment:
I'm not sure I follow why this is helpful. Won't we always iterate the whole list looking for a wild cards ? 

i.e. 
no wildcards -> iterate the whole list on construction, this returns false
leading wildcard -> find the first one in the constructor, then iterate from the second on here. 
trailing wildcard -> find the last one in the constructor, this returns false. 

Isn't this always O(N) regardless if we break up the loop into 2 parts or not ?  Since we are already re-modeling this data and looping in the constructor (effectively transforming from List<String> Iterable<Expression> ) can we simplify this a just transforming one data structure to another in the constuctor ?

Reply:
I have pushed https://github.com/elastic/elasticsearch/pull/92518/commits/2f4aa15b05e7dd47f9eabea4c5246aa48322bab5 which doesn't incur this logic that you're bringing up for discussion here. It still does one `isWildcard` check per list item in the expression (there's no way around it) but it does so eagerly by building another list.